std/smtp/auth.ts

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+// Based on https://github.com/golang/go/blob/92c732e901a732855f4b813e6676264421eceae9/src/net/smtp/auth.go
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+import { HmacSha256 } from "../hash/sha256.ts";
+
+export interface AuthMessage {
+  protocol: string;
+  toServer: string | null;
+}
+
+/** `Auth` is implemented by an SMTP authentication mechanism.
+ */
+export interface Auth {
+  // `start` begins an authentication with a server.
+  // It returns the name of the authentication protocol
+  // and optionally data to include in the initial AUTH message
+  // sent to the server. It can return proto == "" to indicate
+  // that the authentication should be skipped.
+  // If it throw an error, the SMTP client aborts
+  // the authentication attempt and closes the connection.
+  start(server: ServerInfo): AuthMessage;
+
+  // `next` continues the authentication. The server has just sent
+  // the `fromServer` data. If `more` is true, the server expects a
+  // response, which `next` should return as toServer; otherwise
+  // `next` should return toServer == nil.
+  // If `next` returns a non-nil error, the SMTP client aborts
+  // the authentication attempt and closes the connection.
+  next(fromServer: string, more: boolean): string | null;
+}
+
+/** `ServerInfo` records information about an SMTP server.
+ */
+export interface ServerInfo {
+  /** SMTP server name
+   */
+  name: string;
+  /** using TLS, with valid certificate for Name
+   */
+  tls?: boolean | null;
+  /** advertised authentication mechanisms
+   */
+  auth: string[] | null;
+}
+
+class PlainAuth implements Auth {
+  readonly #identity: string;
+  readonly #username: string;
+  readonly #password: string;
+  readonly #host: string;
+  constructor(
+    identity: string,
+    username: string,
+    password: string,
+    host: string,
+  ) {
+    this.#identity = identity;
+    this.#username = username;
+    this.#password = password;
+    this.#host = host;
+  }
+
+  start(server: ServerInfo): AuthMessage {
+    // Must have TLS, or else localhost server.
+    // Note: If TLS is not true, then we can't trust ANYTHING in ServerInfo.
+    // In particular, it doesn't matter if the server advertises PLAIN auth.
+    // That might just be the attacker saying
+    // "it's ok, you can trust me with your password."
+    if (!server.tls && !isLocalhost(server.name)) {
+      throw new Error("unencrypted connection");
+    }
+    if (server.name !== this.#host) {
+      throw new Error("wrong host name");
```

Comment:
Suggestion: Improve logging: `Wrong host name: expected ${this.#host} received ${server.name}`